src/main/java/org/broadinstitute/dsm/util/ElasticSearchUtil.java

```diff
@@ -1405,26 +1348,52 @@ private static void valueQueryBuilder(@NonNull BoolQueryBuilder finalQuery, @Non
         if (wildCard) {
             if (must) {
                 finalQuery.must(QueryBuilders.wildcardQuery(name, "*" + query.toLowerCase() + "*"));
-            }
-            else {
+            } else {
                 finalQuery.should(QueryBuilders.wildcardQuery(name, "*" + query.toLowerCase() + "*"));
             }
-        }
-        else {
+        } else {
             if (must) {
-                finalQuery.must(QueryBuilders.matchQuery(name, query));
-            }
-            else {
+                String fieldType = getFieldTypeByFieldName(name);
+                String finalFieldName = getFinalFieldName(name, fieldType);
+                finalQuery.must(QueryBuilders.termQuery(finalFieldName, query));
+            } else {
                 finalQuery.should(QueryBuilders.matchQuery(name, query));
             }
         }
     }
 
+    private static String getFinalFieldName(String fieldName, String fieldType) {
+        return fieldType.equals(TEXT) ? new StringBuilder(fieldName).append(KEYWORD).toString() : fieldName;
+    }
+
+    private static String getFieldsAsString(String anyStudy) {
+        return fieldMappings.get(anyStudy).source().string();
+    }
+
+    private static String getAnyStudy() {
+        return fieldMappings.keySet().stream().findAny().orElseThrow(() -> new RuntimeException("Error while getting study mapping from ES"));
+    }
+
+
+    private static String getFieldTypeByFieldName(String name) {
+        String anyStudy = getAnyStudy();
+        String fields = getFieldsAsString(anyStudy);
+        String[] fieldsArray = name.split(DOT_SEPARATOR);
+        String outerField = fieldsArray[OUTER_FIELD_INDEX];
+        String innerField = fieldsArray[INNER_FIELD_INDEX];
+        Gson gson = new Gson();
+        HashMap fieldsMap = gson.fromJson(fields, HashMap.class);
```

Comment:
maybe having `Map` instead of `HashMap` on variable declaration would be good with types.

Reply:
Unfortunately the casting mechanism does not work with LinkedTreeMap of Gson, I'm afraid I'll leave this as it is.